README.md
-- SUMMARY --

The Creative Commons field module defines a field type for attaching Creative
Commons licence types. This allows you to add Creative Commons licences to any 
entity type, including media.

For a full description of the module, visit the project page:
  https://www.drupal.org/project/creative_commons

To submit bug reports and feature suggestions, or to track changes:
  https://www.drupal.org/project/issues/creative_commons


-- REQUIREMENTS --

None.


-- INSTALLATION --

* Install as usual, see http://drupal.org/node/1897420 for further information.


-- CONFIGURATION --

* Configure Creative Commons general settings in Administration » Configuration 
» Content authoring » Creative Commons menu 
  (admin/config/content/creative_commons).

* Add a new field in your entity type.


-- CUSTOMIZATION --

* To override default template copy templates/creative-commons.html.twig file
  in your theme. See inside for an overview of the variables you can use.

* To override field template copy 
  templates/field--node--field-creativecommons.html.twig file in your theme and
  change name with your field name. See inside for an overview of the variables 
  you can use.

-- IMAGES --

* Image files from https://licensebuttons.net/i/.

-- CONTACT --

Current maintainer:
* Rafa Morales (rafaticarte) - https://www.drupal.org/user/1186094


CHANGELOG.txt
Creative Commons field 8.x-1.0-beta1, xxxx-xx-xx
----------------------------------
by rafaticarte: Added settings form and field type.
by rafaticarte: Added support for Creative Commons 1.0, 2.0, 2.5, 3.0 and 4.0.
by rafaticarte: Added support for Creative Commons Zero 1.0.
by rafaticarte: Added support for local images and licensebuttons.net images.
by rafaticarte: Added support for overriding template.
by rafaticarte: Added support for views.


tests/src/Functional/LoadTest.php
<?php

namespace Drupal\Tests\creative_commons\Functional;

use Drupal\Core\Url;
use Drupal\Tests\BrowserTestBase;

/**
 * Simple test to ensure that main page loads with module enabled.
 *
 * @group creative_commons
 */
class LoadTest extends BrowserTestBase {

  /**
   * Modules to enable.
   *
   * @var array
   */
  public static $modules = ['creative_commons'];

  /**
   * A user with permission to administer site configuration.
   *
   * @var \Drupal\user\UserInterface
   */
  protected $user;

  /**
   * {@inheritdoc}
   */
  protected function setUp() {
    parent::setUp();
    $this->user = $this->drupalCreateUser(['administer site configuration']);
    $this->drupalLogin($this->user);
  }

  /**
   * Tests that the home page loads with a 200 response.
   */
  public function testLoad() {
    $this->drupalGet(Url::fromRoute('<front>'));
    $this->assertSession()->statusCodeEquals(200);
  }

}


src/Form/SettingsForm.php
<?php

namespace Drupal\creative_commons\Form;

use Drupal\Core\Form\ConfigFormBase;
use Drupal\Core\Form\FormStateInterface;
use Drupal\creative_commons\Repository\CreativeCommonsRepository;

/**
 * Provides a form for setting module.
 */
class SettingsForm extends ConfigFormBase {

  /**
   * {@inheritdoc}
   */
  protected function getEditableConfigNames() {
    return [
      'creative_commons.settings',
    ];
  }

  /**
   * {@inheritdoc}
   */
  public function getFormId() {
    return 'creative_commons_settings_form';
  }

  /**
   * {@inheritdoc}
   */
  public function buildForm(array $form, FormStateInterface $form_state) {
    $form = parent::buildForm($form, $form_state);

    $config = $this->config('creative_commons.settings');
    $creativeCommonsRepository = CreativeCommonsRepository::getInstance();

    $form['versions'] = [
      '#type' => 'checkboxes',
      '#title' => $this->t('License versions available'),
      '#description' => $this->t('Creative Commons license versions available to be choosen by users.'),
      '#options' => $creativeCommonsRepository->getVersions(),
      '#default_value' => $config->get('versions'),
    ];

    $form['image_repository'] = [
      '#type' => 'select',
      '#title' => $this->t('Image repository'),
      '#description' => $this->t('Repository used to load image licenses.'),
      '#options' => [
        'local' => $this->t('Module folder'),
        'licensebuttons' => $this->t('Licensebuttons.net'),
      ],
      '#size' => 1,
      '#default_value' => $config->get('image_repository'),
    ];

    return $form;
  }

  /**
   * {@inheritdoc}
   */
  public function submitForm(array &$form, FormStateInterface $form_state) {
    parent::submitForm($form, $form_state);

    $this->config('creative_commons.settings')
      ->set('versions', $form_state->getValue('versions'))
      ->set('image_repository', $form_state->getValue('image_repository'))
      ->save();
  }

}


src/Plugin/Field/FieldType/CreativeCommonsItem.php
<?php

namespace Drupal\creative_commons\Plugin\Field\FieldType;

use Drupal\Core\Field\FieldItemBase;
use Drupal\Core\Field\FieldStorageDefinitionInterface;
use Drupal\Core\StringTranslation\TranslatableMarkup;
use Drupal\Core\TypedData\DataDefinition;

/**
 * Plugin implementation of the 'creative_commons' field type.
 *
 * @FieldType(
 *   id = "creative_commons",
 *   label = @Translation("Creative Commons"),
 *   description = @Translation("This field stores a Creative Commons license and its data in the database."),
 *   default_widget = "creative_commons",
 *   default_formatter = "creative_commons"
 * )
 */
class CreativeCommonsItem extends FieldItemBase {

  /**
   * {@inheritdoc}
   */
  public static function defaultStorageSettings() {
    return [
      'max_length' => 255,
      'max_length_id' => 20,
      'is_ascii' => FALSE,
      'case_sensitive' => FALSE,
    ] + parent::defaultStorageSettings();
  }

  /**
   * {@inheritdoc}
   */
  public static function schema(FieldStorageDefinitionInterface $field_definition) {
    $schema = [
      'columns' => [
        'cc_id' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length_id'),
          'not null' => FALSE,
        ],
        'work_title' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
        'work_link' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
        'author_name' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
        'author_link' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
        'source_title' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
        'source_link' => [
          'type' => 'varchar',
          'length' => (int) $field_definition->getSetting('max_length'),
          'not null' => FALSE,
        ],
      ],
    ];

    return $schema;
  }

  /**
   * {@inheritdoc}
   */
  public function isEmpty() {
    $value = $this->get('cc_id')->getValue();
    return $value === NULL || $value === '';
  }

  /**
   * {@inheritdoc}
   */
  public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition) {
    $properties['cc_id'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Creative Commons license ID'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['work_title'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Work title'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['work_link'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Work link'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['author_name'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Author name'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['author_link'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Author link'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['source_title'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Source work title'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    $properties['source_link'] = DataDefinition::create('string')
      ->setLabel(new TranslatableMarkup('Source work link'))
      ->setSetting('case_sensitive', $field_definition->getSetting('case_sensitive'))
      ->setRequired(FALSE);

    return $properties;
  }

}


src/Plugin/Field/FieldWidget/CreativeCommonsWidget.php
<?php

namespace Drupal\creative_commons\Plugin\Field\FieldWidget;

use Drupal\Core\Config\ConfigFactoryInterface;
use Drupal\Core\Field\FieldDefinitionInterface;
use Drupal\Core\Field\FieldItemListInterface;
use Drupal\Core\Field\WidgetBase;
use Drupal\Core\Form\FormStateInterface;
use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
use Drupal\creative_commons\Repository\CreativeCommonsRepository;
use Symfony\Component\DependencyInjection\ContainerInterface;

/**
 * Plugin implementation of the 'creative_commons' widget.
 *
 * @FieldWidget(
 *   id = "creative_commons",
 *   module = "creative_commons",
 *   label = @Translation("Creative Commons"),
 *   field_types = {
 *     "creative_commons"
 *   }
 * )
 */
class CreativeCommonsWidget extends WidgetBase implements ContainerFactoryPluginInterface {
  /**
   * The configuration factory.
   *
   * @var \Drupal\Core\Config\ConfigFactoryInterface
   */
  protected $configFactory;

  /**
   * {@inheritdoc}
   */
  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, array $third_party_settings, ConfigFactoryInterface $config_factory) {
    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $third_party_settings);
    $this->configFactory = $config_factory;
  }

  /**
   * {@inheritdoc}
   */
  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
    return new static(
      $plugin_id,
      $plugin_definition,
      $configuration['field_definition'],
      $configuration['settings'],
      $configuration['third_party_settings'],
      $container->get('config.factory')
    );
  }

  /**
   * {@inheritdoc}
   */
  public static function defaultSettings() {
    return [
      'size' => 60,
      'max_length' => 255,
    ] + parent::defaultSettings();
  }

  /**
   * {@inheritdoc}
   */
  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
    $config = $this->configFactory->get('creative_commons.settings');
    $creativeCommonsRepository = CreativeCommonsRepository::getInstance();

    $element = [
      '#type' => 'details',
      '#title' => $items->getFieldDefinition()->getLabel(),
      '#open' => TRUE,
      '#element_validate' => [
        [$this, 'validateForm'],
      ],
    ];

    $element['cc_id'] = [
      '#type' => 'select',
      '#title' => $this->t('Creative Commons license type'),
      '#default_value' => isset($items[$delta]->cc_id) ? $items[$delta]->cc_id : NULL,
      '#required' => FALSE,
      '#options' => $creativeCommonsRepository->getLicenses($config->get('versions')),
    ];

    $element['work_title'] = [
      '#type' => 'textfield',
      '#title' => $this->t('Work title'),
      '#description' => $this->t('Leave blank to show the title of content. Only available in <em>node</em> entities.'),
      '#default_value' => isset($items[$delta]->work_title) ? $items[$delta]->work_title : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    $element['work_link'] = [
      '#type' => 'textfield',
      '#title' => $this->t('Work link'),
      '#description' => $this->t('Leave blank to link content. Only available in <em>node</em> entities.'),
      '#default_value' => isset($items[$delta]->work_link) ? $items[$delta]->work_link : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    $element['author_name'] = [
      '#type' => 'textfield',
      '#title' => $this->t('Author name'),
      '#description' => $this->t('Leave blank to show the author name of content. Only available in <em>node</em> entities.'),
      '#default_value' => isset($items[$delta]->author_name) ? $items[$delta]->author_name : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    $element['author_link'] = [
      '#type' => 'url',
      '#title' => $this->t('Author link'),
      '#description' => $this->t('Leave blank to link the author of content. Only available in <em>node</em> entities.'),
      '#default_value' => isset($items[$delta]->author_link) ? $items[$delta]->author_link : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    $element['source_title'] = [
      '#type' => 'textfield',
      '#title' => $this->t('Source work title'),
      '#description' => $this->t('Title of work in which it is based on.'),
      '#default_value' => isset($items[$delta]->source_title) ? $items[$delta]->source_title : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    $element['source_link'] = [
      '#type' => 'url',
      '#title' => $this->t('Source work link'),
      '#description' => $this->t('Link of work in which it is based on.'),
      '#default_value' => isset($items[$delta]->source_link) ? $items[$delta]->source_link : NULL,
      '#size' => $this->getSetting('size'),
      '#maxlength' => $this->getSetting('max_length'),
      '#required' => FALSE,
    ];

    return $element;
  }

  /**
   * Validate form elements and show errors.
   *
   * This function is assigned as an #element_validate
   * callback in formElement().
   */
  public function validateForm($element, FormStateInterface $form_state) {
    if ($element['work_title']['#value'] != strip_tags($element['work_title']['#value'])) {
      $form_state->setError($element['work_title'], $this->t('No HTML tags allowed.'));
    }
    if ($element['author_name']['#value'] != strip_tags($element['author_name']['#value'])) {
      $form_state->setError($element['author_name'], $this->t('No HTML tags allowed.'));
    }
    if ($element['source_title']['#value'] != strip_tags($element['source_title']['#value'])) {
      $form_state->setError($element['source_title'], $this->t('No HTML tags allowed.'));
    }
    if ($element['author_link']['#value'] != '' && $element['author_name']['#value'] == '') {
      $form_state->setError($element['author_name'], $this->t('Author name cannot be blank if you fill in author link.'));
    }
    if ($element['source_link']['#value'] != '' && $element['source_title']['#value'] == '') {
      $form_state->setError($element['source_title'], $this->t('Source work title cannot be blank if you fill in source work link.'));
    }
  }

}


src/Plugin/Field/FieldFormatter/CreativeCommonsFormatter.php
<?php

namespace Drupal\creative_commons\Plugin\Field\FieldFormatter;

use Drupal\Core\Config\ConfigFactoryInterface;
use Drupal\Core\Field\FieldDefinitionInterface;
use Drupal\Core\Field\FieldItemListInterface;
use Drupal\Core\Field\FormatterBase;
use Drupal\Core\Form\FormStateInterface;
use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
use Drupal\Core\Routing\CurrentRouteMatch;
use Drupal\creative_commons\Repository\CreativeCommonsRepository;
use Symfony\Component\DependencyInjection\ContainerInterface;

/**
 * Plugin implementation of the 'creative_commons' formatter.
 *
 * @FieldFormatter(
 *   id = "creative_commons",
 *   label = @Translation("Creative Commons"),
 *   field_types = {
 *     "creative_commons"
 *   }
 * )
 */
class CreativeCommonsFormatter extends FormatterBase implements ContainerFactoryPluginInterface {

  /**
   * The configuration factory.
   *
   * @var \Drupal\Core\Config\ConfigFactoryInterface
   */
  protected $configFactory;

  /**
   * The configuration factory.
   *
   * @var \Drupal\Core\Routing\CurrentRouteMatch
   */
  protected $routeMatch;

  /**
   * {@inheritdoc}
   */
  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, ConfigFactoryInterface $config_factory, CurrentRouteMatch $route_match) {
    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings);
    $this->configFactory = $config_factory;
    $this->routeMatch = $route_match;
  }

  /**
   * {@inheritdoc}
   */
  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
    return new static(
      $plugin_id,
      $plugin_definition,
      $configuration['field_definition'],
      $configuration['settings'],
      $configuration['label'],
      $configuration['view_mode'],
      $configuration['third_party_settings'],
      $container->get('config.factory'),
      $container->get('current_route_match')
    );
  }

  /**
   * {@inheritdoc}
   */
  public static function defaultSettings() {
    return [
      'license_name' => 'long',
      'image_position' => 'above',
      'image_size' => 'medium',
    ] + parent::defaultSettings();
  }

  /**
   * {@inheritdoc}
   */
  public function settingsForm(array $form, FormStateInterface $form_state) {
    $elements['license_name'] = [
      '#title' => $this->t('License name'),
      '#type' => 'select',
      '#options' => [
        'long' => $this->t('Long'),
        'short' => $this->t('Short'),
      ],
      '#default_value' => $this->getSetting('license_name'),
    ];

    $elements['image_position'] = [
      '#title' => $this->t('Image position'),
      '#type' => 'select',
      '#options' => [
        'above' => $this->t('Above'),
        'inline' => $this->t('Inline'),
        'hidden' => $this->t('Hidden'),
      ],
      '#default_value' => $this->getSetting('image_position'),
    ];

    $elements['image_size'] = [
      '#title' => $this->t('Image size'),
      '#type' => 'select',
      '#options' => [
        'medium' => $this->t('Medium'),
        'small' => $this->t('Small'),
      ],
      '#default_value' => $this->getSetting('image_size'),
    ];

    return $elements;
  }

  /**
   * {@inheritdoc}
   */
  public function settingsSummary() {
    $summary = [];

    $summary[] = $this->t('License name: @license_name', ['@license_name' => ucfirst($this->getSetting('license_name'))]);
    $summary[] = $this->t('License image position: @image_position', ['@image_position' => ucfirst($this->getSetting('image_position'))]);
    $summary[] = $this->t('License image size: @image_size', ['@image_size' => ucfirst($this->getSetting('image_size'))]);

    return $summary;
  }

  /**
   * {@inheritdoc}
   */
  public function viewElements(FieldItemListInterface $items, $langcode) {
    $config = $this->configFactory->get('creative_commons.settings');
    $creativeCommonsRepository = CreativeCommonsRepository::getInstance();
    $entity = $items->getEntity();

    $elements = [];

    foreach ($items as $delta => $item) {
      if ($item->work_title == '' and $entity->getEntityTypeId() == 'node') {
        $work_title = $entity->getTitle();
        $work_link = $entity->toURL()->toString();
      }
      else {
        $work_title = $item->work_title;
        $work_link = $item->work_link;
      }

      if ($item->author_name == '' and $entity->getEntityTypeId() == 'node') {
        $author_name = $entity->getOwner()->getDisplayName();
        $author_link = $entity->getOwner()->toURL()->toString();
      }
      elseif ($creativeCommonsRepository->isZero($item->cc_id)) {
        $author_name = '';
        $author_link = '';
      }
      else {
        $author_name = $item->author_name;
        $author_link = $item->author_link;
      }

      $elements[$delta] = [
        '#theme' => 'creative_commons',
        '#image_position' => $this->getSetting('image_position'),
        '#cc_image' => $creativeCommonsRepository->getImage($item->cc_id, $this->getSetting('image_size'), $config->get('image_repository')),
        '#cc_name' => $creativeCommonsRepository->getName($item->cc_id, $this->getSetting('license_name')),
        '#cc_legal' => $creativeCommonsRepository->getLegal($item->cc_id),
        '#work_title' => $work_title,
        '#work_link' => $work_link,
        '#author_name' => $author_name,
        '#author_link' => $author_link,
        '#source_title' => $item->source_title,
        '#source_link' => $item->source_link,
      ];
    }

    return $elements;
  }

}


src/Repository/CreativeCommonsRepository.php
<?php

namespace Drupal\creative_commons\Repository;

use Drupal\Core\StringTranslation\StringTranslationTrait;

/**
 * Defines a repository for storage Creative Commons licenses.
 */
class CreativeCommonsRepository {
  use StringTranslationTrait;

  /**
   * Object for Singleton instance.
   *
   * @var instance
   */
  private static $instance;

  /**
   * Array with the content of CreativeCommonsLicense.json file.
   *
   * @var repository
   */
  private $repository;

  /**
   * Construct for loading CreativeCommonsLicense.json file.
   */
  private function __construct() {
    $file = file_get_contents(drupal_get_path('module', 'creative_commons') . '/includes/CreativeCommonsLicenses.json');
    if ($file == FALSE) {
      \Drupal::logger('creative_commons')->error('<em>CreativeCommonsLicenses.json</em> file is not found');
      $this->repository = [];
    }
    else {
      $this->repository = json_decode($file, TRUE);
      if ($this->repository == NULL) {
        \Drupal::logger('creative_commons')->error('<em>CreativeCommonsLicenses.json</em> file is not JSON valid');
        $this->repository = [];
      }
    }
  }

  /**
   * Singleton pattern.
   *
   * @return object
   *   Instance of CreativeCommonsRepository.
   */
  public static function getInstance() {
    if (!self::$instance instanceof self) {
      self::$instance = new self();
    }
    return self::$instance;
  }

  /**
   * Get translation name of Creative Commons license.
   *
   * @param string $cc_id
   *   Identifier of license.
   * @param string $name_length
   *   Length of license name.
   *
   * @return string
   *   Translation name of license.
   */
  public function getName($cc_id, $name_length = 'long') {
    if (empty($this->repository)) {
      return '';
    }

    if ($name_length == 'short') {
      $name = 'CC ';
      $name .= strtoupper($this->repository['types'][$cc_id]['terms']) . ' ';
      $name .= $this->repository['types'][$cc_id]['version'];
    }
    else {
      $name = $this->t('Creative Commons') . ' ';

      switch ($this->repository['types'][$cc_id]['terms']) {
        case 'by':
          $name .= $this->t('Attribution') . ' ';
          break;

        case 'by-nc':
          $name .= $this->t('Attribution') . '-' . $this->t('NonCommercial') . ' ';
          break;

        case 'by-nc-nd':
          $name .= $this->t('Attribution') . '-' . $this->t('NonCommercial') . '-' . $this->t('NoDerivatives') . ' ';
          break;

        case 'by-nc-sa':
          $name .= $this->t('Attribution') . '-' . $this->t('NonCommercial') . '-' . $this->t('ShareAlike') . ' ';
          break;

        case 'by-nd':
          $name .= $this->t('Attribution') . '-' . $this->t('NoDerivatives') . ' ';
          break;

        case 'by-sa':
          $name .= $this->t('Attribution') . '-' . $this->t('ShareAlike') . ' ';
          break;

        case 'zero':
          $name .= $this->t('Zero') . ' ';
          break;
      }

      switch ($this->repository['types'][$cc_id]['version']) {
        case '4.0':
          $name .= $this->repository['types'][$cc_id]['version'] . ' ' . $this->t('International');
          break;

        case '3.0':
          $name .= $this->repository['types'][$cc_id]['version'] . ' ' . $this->t('Unported');
          break;

        case '2.5':
        case '2.0':
        case '1.0':
          $name .= $this->repository['types'][$cc_id]['version'] . ' ' . $this->t('Generic');
          break;

        case 'zero':
          $name .= $this->repository['types'][$cc_id]['version'] . ' ' . $this->t('Universal');
          break;
      }
    }

    return $name;
  }

  /**
   * Get link to image license.
   *
   * @param string $cc_id
   *   Identifier of license.
   * @param string $size
   *   Size of image license.
   * @param string $source
   *   Source of image license.
   *
   * @return string
   *   Link to image license.
   */
  public function getImage($cc_id, $size = 'medium', $source = 'local') {
    if (empty($this->repository)) {
      return '';
    }

    global $base_path;
    $terms = $this->repository['types'][$cc_id]['terms'];
    $version = $this->repository['types'][$cc_id]['version'];

    $image = '';

    switch ($source) {
      case 'local':
        $image .= $base_path . drupal_get_path('module', 'creative_commons') . '/images/';
        break;

      case 'licensebuttons':
        $image .= 'https://licensebuttons.net/l/';
        break;
    }

    $image .= $terms . '/' . $version . '/';

    switch ($size) {
      case 'small':
        $image .= '80x15.png';
        break;

      case 'medium':
        $image .= '88x31.png';
        break;
    }

    return $image;
  }

  /**
   * Get link to legal information license.
   *
   * @param string $cc_id
   *   Identifier of license.
   *
   * @return string
   *   Link to legal information license.
   */
  public function getLegal($cc_id) {
    if (empty($this->repository)) {
      return '';
    }

    $terms = $this->repository['types'][$cc_id]['terms'];
    $version = $this->repository['types'][$cc_id]['version'];

    return 'https://creativecommons.org/licenses/' . $terms . '/' . $version . '/';
  }

  /**
   * Check if a license is Zero type.
   *
   * @param string $cc_id
   *   Identifier of license.
   *
   * @return bool
   *   TRUE if license is Zero type.
   */
  public function isZero($cc_id) {
    if ($this->repository['types'][$cc_id]['terms'] == 'zero') {
      return TRUE;
    }
    else {
      return FALSE;
    }
  }

  /**
   * Defines ....
   *
   * $config->get('versions') viene con valores a 0
   */
  public function getLicenses($versions = []) {
    $licenses = ['' => 'None'];

    if (empty($this->repository)) {
      return $licenses;
    }

    foreach ($versions as $cc_version) {
      if ($cc_version != '0') {
        $licenses[$this->repository['versions'][$cc_version]['title']] = [];
        foreach ($this->repository['versions'][$cc_version]['types'] as $item) {
          $licenses[$this->repository['versions'][$cc_version]['title']] += [$item => $this->getName($item)];
        }
      }
    }

    return $licenses;
  }

  /**
   * Get Creative Commons versions available.
   *
   * @return array
   *   Array with Creative Commons versions id and name.
   */
  public function getVersions() {
    $versions = [];

    foreach ($this->repository['versions'] as $cc_version_id => $cc_version_data) {
      $versions[$cc_version_id] = $cc_version_data['title'];
    }

    return $versions;
  }

}
